### PR TITLE
fix: improve slots mounting option

### LINKED ISSUE #793: $parent of a child component defined on a slot, reference to itself
### Version
1.0.0-beta.20

### Reproduction link
[https://codesandbox.io/s/wqo716rm55](https://codesandbox.io/s/wqo716rm55)

### Steps to reproduce
Run the test written on sandbox

### What is expected?
It should work as works on the browser so `this.$parent` should return the parent of the component and not itself.

### What is actually happening?
I have a scenario similar to what I reproduce in the sandbox. I have a few components that are inserted on the default slot of a "wrapper" component, each of these components are registered on the parent when are mounted using a method on the parent `this.$parent.register`.
All work on the browser but when I tried to test the `this.$parent` of the child component refers to itself and the test fails because it doesn't find the register method.

---
For example If I change `this.$parent` by `this.$parent.$parent` all works.

### PR DESCRIPTION
**Pull request by 38elements**
This resolves #793.
This is related to #813.
This does not reslove #827.

### CODE REVIEW

**Comment by 38elements** [comment]
I will add some tests and improvements within 3 days.

File: packages/create-instance/create-render-slot.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import { compileToFunctions } from 'vue-template-compiler'
+
+const _renderSlot = Vue.prototype._t
+
+function createVNodes (
+  vm: Component,
+  slotValue: Component | string
+): ?Array<VNode> {
+  if (typeof slotValue === 'string') {
+    const compiledResult = compileToFunctions(`<div>${slotValue}{{ }}</div>`)
```
Comment (eddyerburgh) [inline packages/create-instance/create-render-slot.js]: Why the empty interpolation `{{ }}`?

Comment (38elements) [reply packages/create-instance/create-render-slot.js]: I do not understand.
It is not necessary.
I thought the compileToFunctions raises an error if it dose not exist at #274.

**Comment by 38elements** [comment]
I think it is necessary to add warning if  `compileToFunctions` is not existing.
I will add them.

File: packages/create-instance/create-render-slot.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+function createVNodes (
+  vm: Component,
+  slotValue: Component | string
+): ?Array<VNode> {
+  if (typeof slotValue === 'string') {
+    // Since compileToFunctions is checked in createSlotVNodes(),
+    // it is not necessary to check compileToFunctions.
+    const compiledResult = compileToFunctions(`<div>${slotValue}</div>`)
+    const _staticRenderFns = vm._renderProxy.$options.staticRenderFns
+    vm._renderProxy.$options.staticRenderFns = compiledResult.staticRenderFns
```
Comment (eddyerburgh) [inline packages/create-instance/create-render-slot.js:17]: Why do you save the renderFunctions and then reassign?

File: test/specs/mounting-options/slots.spec.js [hunk 3]
```diff
@@ -568,4 +574,72 @@ describeWithMountingMethods('options.slots', mountingMethod => {
...
     }
   })
+
+  itDoNotRunIf(
+    mountingMethod.name === 'renderToString',
+    'mounts component with default slot if passed string in slot object',
+    () => {
+      const wrapper1 = mount(ComponentWithSlots, { slots: { default: 'foo<span>123</span>{{ foo }}' }})
+      expect(wrapper1.find('main').html()).to.equal('<main>foo<span>123</span>bar</main>')
+      const wrapper2 = mount(ComponentWithSlots, { slots: { default: '<p>1</p>{{ foo }}2' }})
```
Comment (eddyerburgh) [inline test/specs/mounting-options/slots.spec.js:584]: I don't think this is how slots should behave. foo should not be resolved using the child vm that it's rendered in. If it uses any instance to resolve foo on, it should be the root instance. But I don't think we should support any variables in the slots option as it adds too much complexity

File: packages/create-instance/create-render-slot.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    const vnodes = compiledResult.render.call(
+      vm._renderProxy, vm.$createElement
+    ).children
+    vm._renderProxy.$options.staticRenderFns = _staticRenderFns
+    return vnodes
+  }
+  return [vm.$createElement(slotValue)]
+}
+
+export default function createRenderSlot (
```
Comment (eddyerburgh) [inline packages/create-instance/create-render-slot.js:27]: What is the purpose of replacing the renderSlot alias? Does this solve the issue of parent being undefined?

File: test/specs/mounting-options/slots.spec.js (hunk 3 above)
Comment (38elements) [reply test/specs/mounting-options/slots.spec.js:584]: I am sorry.
Your pointing out is correct.
I misunderstood about it.

**Pull request closed without merging**